Handle resolution errors in @wordpress/data

adamziel opened the pull request:
## Description

`@wodrpress/data` exposes a resolvers API, but ignores any errors thrown during the resolution:

https://github.com/WordPress/gutenberg/blob/70730e61ba3148ebfa2be125d17c8138ab16392c/packages/core-data/src/resolvers.js#L107-L111

This PR proposes an error handling mechanism that works roughly as follows:

* Any thrown exception is stored using the new `failResolution` action which also finalizes the resolution
* The error details may be fetched using the `hasLastResolutionFailed` and `getLastResolutionFailure` selectors
* A promise returned by the `resolveSelect` will now be rejected in case of resolution error. Previously it just hanged forever.

To that last point, we may need to wrap any `resolveSelect` in try/catch to make sure redux actions don't start unexpectedly throwing exceptions.

## Test plan

It is a new feature, so testing is mostly about confirming the CI checks, reading the code, and discussing.

cc @youknowriad @jsnajdr @mcsf @gziolo @ellatrix @draganescu @noisysocks @jsnajdr @Mamaduka @jorgefilipecosta @talldan @getdave @kevin940726 @anton-vlasenko

adamziel committed:
First stab at the FAIL_RESOLUTION meta action

adamziel committed:
Adjust hasStartedResolution

kevin940726 commented:
Haven't reviewed anything yet. Just want to say that you're a superhero, @adamziel 😍 . What would we do without you?

adamziel committed:
Remove try/catch to see if it helps with e2e failures

adamziel committed:
Adjust resolvers-cache-middleware to use the new state shape

adamziel committed:
Add unit tests

adamziel committed:
Improve test coverage

adamziel committed:
make resolveSelect reject on failure

adamziel committed:
Fix unit tests

youknowriad commented on packages/data/src/redux-store/metadata/selectors.js:
```diff
@@ -54,6 +71,36 @@ export function hasFinishedResolution( state, selectorName, args = [] ) {
+ *
+ * @return {boolean} Has resolution failed
+ */
+export function hasLastResolutionFailed( state, selectorName, args = [] ) {
```
Should we drop the "last" from the name of these selectors. I mean I see the value in clarity but it's a departure from isResolving which is also about the "last resolution"

adamziel replied:
Great idea @youknowriad! I just updated the PR. I borrowed this naming from `getLastEntityDeleteError` and `getLastEntitySaveError` in `core-data`. I wonder if we should deprecate them in favor of `getEntityDeleteError` and `getEntitySaveError`? Either way, that's for another PR.

adamziel committed:
Replace LastResolution with just Resolution

Mamaduka commented on packages/data/src/redux-store/index.js:
```diff
@@ -333,20 +333,35 @@ function mapResolveSelectors( selectors, store ) {
 				const hasFinished = () =>
 					selectors.hasFinishedResolution( selectorName, args );
+				const finalize = ( result ) => {
+					const hasFailed = selectors.hasResolutionFailed(
+						selectorName,
+						args
+					);
+					if ( hasFailed ) {
+						const error = selectors.getResolutionFailure(
+							selectorName,
+							args
+						);
```
**Suggestion (non-blocking):** I think we can  simplify this by checking if there's an error and rejecting.

```js
const error = selectors.getResolutionFailure(
  selectorName,
  args
);

if ( error ) {
  reject( error );
}
```

adamziel replied:
As much as I'd like that, an error could also be `null` or `undefined`, e.g. consider this snippet:

```js
try {
  throw null;
} catch(e) {
  console.log(e);
}
```

jsnajdr commented on packages/data/src/redux-store/metadata/selectors.js:
```diff
@@ -54,6 +71,36 @@ export function hasFinishedResolution( state, selectorName, args = [] ) {
+ *
+ * @return {Error|undefined} Last resolution error
+ */
+export function getResolutionFailure( state, selectorName, args = [] ) {
```
I'd call this `getResolutionError`.

jsnajdr commented on packages/data/src/redux-store/metadata/selectors.js:
```diff
@@ -54,6 +71,36 @@ export function hasFinishedResolution( state, selectorName, args = [] ) {
+ *
+ * @return {boolean} Has resolution failed
+ */
+export function hasResolutionFailed( state, selectorName, args = [] ) {
```
Does it make sense to default `args` to `[]`? The existing `getIsResolving` selector doesn't do that. That's at least inconsistent: some selectors do default, other don't.

adamziel replied:
I agree it's inconsistent. It follows the other consumers of `getIsResolving` like `hasFinishedResolution`, `isResolving`, and `hasStartedResolution`. That empty list is unfortunately required to retrieve the correct map entry, otherwise the consumers of this selector would have to specify that when needed.

That's beyond the scope of this PR, but let's consider how we could make it consistent.

One way would be to store the argument-less resolution state under the `undefined` key. I'd worry this it would break some consumers of this API out in the wild, though. E.g. it would break any code that calls `hasStartedResolution( selectorName, args )` where args has varying length of 0 or more.

The other alternative is to give all the selectors the default argument value of `[]`. It doesn't seem to make sense to explicitly pass undefined as that third parameter so I think it's as backwards compatible as we can get.

jsnajdr replied:
I (independently of this PR) discovered there are even some bugs in how the `args` array is handled and how the map entries are retrieved. I'm fixing that thoroughly in #38945.

adamziel replied:
That's a great one! I just rebased this PR on top of trunk and removed the default `[]` value for arguments.

jsnajdr commented on packages/data/src/redux-store/metadata/selectors.js:
```diff
@@ -54,6 +71,36 @@ export function hasFinishedResolution( state, selectorName, args = [] ) {
+export function hasResolutionFailed( state, selectorName, args = [] ) {
+	const resolutionState = getResolutionState( state, selectorName, args );
+	const hasFailed = resolutionState && 'error' in resolutionState;
+	return !! hasFailed;
```
You could as well return `!! getResolutionFailure( state, selectorName, args )`, it's equivalent.

adamziel replied:
It's not equivalent, we store any `e` caught by `} catch( e ) { ` and that could be falsy, e.g.:

```js
try {
  throw null;
} catch(e) {
  console.log(e);
}
```

jsnajdr commented on packages/data/src/redux-store/metadata/test/selectors.js:
```diff
@@ -12,6 +12,20 @@ import {
+ */
+import { createRegistry } from '@wordpress/data';
+
+jest.useFakeTimers();
```
Would it be possible to write the same tests with `useRealTimers()`? And instead of calling `advanceTimersByTime`, do some `await`-ing where appropriate?

adamziel replied:
good idea, conveniently it just worked without any further changes

jsnajdr commented on packages/data/src/redux-store/metadata/actions.js:
```diff
@@ -68,6 +87,26 @@ export function finishResolutions( selectorName, args ) {
+ *                                      is associated to a resolution.
+ * @param {(Error|null)[]} errors       Array of errors to associate for uniqueness, each item
+ *                                      is associated to a resolution.
+ * @return {{ type: 'FAIL_RESOLUTIONS', selectorName: string, args: unknown[], errors: Array<Error|null> }} Action object.
```
Can there be a failure that has a `null` error? I'd expect that where there is a failure, there is always an error object, or isn't there?

adamziel replied:
Yes, see https://github.com/WordPress/gutenberg/pull/38669#discussion_r809148493

Perhaps this should even be `Array<unknown>`.

adamziel replied:
I went for `Error | unknown` and added a note for consumers of the API inside of the relevant selector's docstring.

jsnajdr commented on packages/data/src/redux-store/metadata/reducer.ts:
```diff
@@ -36,17 +40,41 @@ const subKeysIsResolved: Reducer< Record< string, State >, Action > = onSubKey<
+		}
+		case 'FAIL_RESOLUTIONS': {
+			const nextState = new EquivalentKeyMap( state );
+			for ( const idx in action.args ) {
```
Going through an array with `in` is a weird style 🙂 I'd do:
```js
action.args.forEach( ( resolutionArgs, idx ) => {
  nextState.set( resolutionArgs, { isResolving: false, error: action.errors[ idx ] } );
}
```

adamziel replied:
Makes sense! I used the `forEach` but kept the conditional setting of the `error` key.

jsnajdr commented on packages/data/src/redux-store/metadata/reducer.ts:
```diff
@@ -36,17 +40,41 @@ const subKeysIsResolved: Reducer< Record< string, State >, Action > = onSubKey<
 		case 'START_RESOLUTION':
 		case 'FINISH_RESOLUTION': {
-			const isStarting = action.type === 'START_RESOLUTION';
+			const isResolving = action.type === 'START_RESOLUTION';
```
I'd personally split both cases into separate chunks of code:
```js
case START:
  ...
case FINISH:
  ...
case FAIL:
  ...
```
The `nextState = new ...` and `return nextState` statements are repeated, but the whole thing is easier to read IMO.

adamziel committed:
Rename getResolutionFailure to getResolutionError

adamziel committed:
Use real timers in Jest tests

adamziel committed:
Simplify the reducer code

adamziel committed:
Add a unit test for the falsy error case

adamziel commented:
I added a [test case for falsy errors](https://github.com/WordPress/gutenberg/pull/38669/commits/0c77c84c2e797cb2595b2183a498e34a762b9879):

```js
	it( 'returns true if the resolution has failed even if the error is falsy', async () => {
		registry.registerStore( 'store', {
			reducer: ( state = null, action ) => {
				if ( action.type === 'RECEIVE' ) {
					return action.items;
				}

				return state;
			},
			selectors: {
				getItems: ( state ) => state,
			},
			resolvers: {
				getItems: () => {
					throw null;
				},
			},
		} );

		expect(
			registry.select( 'store' ).hasResolutionFailed( 'getItems' )
		).toBeFalsy();

		await resolve( registry, 'getItems' );

		expect(
			registry.select( 'store' ).hasResolutionFailed( 'getItems' )
		).toBeTruthy();
	} );
```

adamziel committed:
Add the missing selectorArgsToStateKey call

adamziel committed:
Align doc strings

adamziel committed:
Denote that error may be of unknown type in the docstring

adamziel committed:
Explain that an error may be of unknown type

adamziel committed:
Adjust typescript type hints

adamziel committed:
Fix tests, use the actual store to test selectors

adamziel committed:
Remove dev artifact

adamziel commented:
@jsnajdr this one is ready for another review 🙏

kevin940726 commented on packages/data/src/redux-store/metadata/reducer.ts:
```diff
@@ -13,15 +13,19 @@ import { selectorArgsToStateKey, onSubKey } from './utils';
 
-export type State = EquivalentKeyMap< unknown[] | unknown, boolean >;
+type StateKey = unknown[] | unknown;
+export type StateValue = { isResolving: boolean; error?: Error | unknown };
```
I wonder if it makes sense to just use the `status` enum in the reducer rather than using `isResolving` boolean? So that we can keep the state as simple as possible, higher-level selectors can interpret the value and expose `getIsResolving` or `getHasResolved` from there.

adamziel replied:
I like that idea! I just updated the PR, let's hope all the checks pass

adamziel replied:
@kevin940726 Let me also loop in @youknowriad @jsnajdr and @gziolo, it would be great to get this one in

youknowriad replied:
I personally don't have strong opinion on what we do in reducers :P it's just internal implementation detail.

adamziel replied:
@youknowriad I was pretty cryptic, wasn't I? :D I meant to ask for your review of this PR – it would be great to get it shipped.

adamziel committed:
Refactor isResolving to enum-based status

kevin940726 commented on packages/data/src/redux-store/index.js:
```diff
@@ -413,15 +428,28 @@ function mapResolvers( resolvers, selectors, store, resolversCache ) {
+								args
+							)
+						);
+					} catch ( e ) {
```
Should better use readable name 😅:

```suggestion
					} catch ( error ) {
```

kevin940726 commented on packages/data/src/redux-store/metadata/reducer.ts:
```diff
@@ -13,15 +13,26 @@ import { selectorArgsToStateKey, onSubKey } from './utils';
 
-export type State = EquivalentKeyMap< unknown[] | unknown, boolean >;
+type StateKey = unknown[] | unknown;
+export enum Status {
```
We should still have an `IDLE` state even though we just leave it as `undefined` for now. Could be easier to reason about when we transform it into a state machine or something like that in the future.

And in `getResolutionState`, we can just always return `StateValue` if we automatically map `undefined` to a constant `{ status: STATUS.IDLE }` object.

adamziel replied:
> Could be easier to reason about when we transform it into a state machine or something like that in the future.

This may or may not ever happen. I'd say YAGNI. We can always react when that happens. I like your second point, though so let's try that. Enums, unfortunately, don't support undefined values, but maybe a regular type union would work just fine here.

adamziel replied:
Done @kevin940726

kevin940726 replied:
I mean instead of using `undefined` as an enum value, let's just map it to `IDLE` internally. 

```ts
enum StateKey {
  IDLE: 'IDLE',
  RESOLVING: 'RESOLVING',
  // ...
}

// In getResolutionState()
if ( ! resolutionState ) {
    return { status: StatusKey.IDLE }
}
```

So that we don't have to use a random undefined in a union type.

adamziel replied:
I'd rather not do that, the reducer never returns that value. Now that I think about it more, it could add confusion to  `createResolversCacheMiddleware` where we access the raw state and not the value returned by `getResolutionState`. I reverted these types to simply reflect the underlying state without any augmenting – I'm still happy to keep discussing that separately from this PR.

kevin940726 replied:
I don't feel strongly about this anyway 😆, just approved! Thank you for your hard work and patience! 🙇‍♂️

kevin940726 commented on packages/data/src/redux-store/metadata/selectors.js:
```diff
@@ -42,7 +67,9 @@ export function getIsResolving( state, selectorName, args ) {
 export function hasStartedResolution( state, selectorName, args ) {
-	return getIsResolving( state, selectorName, args ) !== undefined;
+	return (
+		getResolutionState( state, selectorName, args )?.status !== undefined
```
Since that `status` will always be defined if the state is defined, I guess we can just do this:

```suggestion
		getResolutionState( state, selectorName, args ) !== undefined
```

adamziel replied:
This hinges on https://github.com/WordPress/gutenberg/pull/38669#discussion_r817348750, I removed the '?' for now

kevin940726 commented on packages/data/src/redux-store/metadata/selectors.js:
```diff
@@ -20,9 +25,9 @@ import { selectorArgsToStateKey } from './utils';
+ * @return {StateValue | undefined} isResolving value.
  */
-export function getIsResolving( state, selectorName, args ) {
+export function getResolutionState( state, selectorName, args ) {
```
Nit: we can also have a user-facing API called `getResolutionStatus` to indicate which `status`(`StatusKey`) the resolution is in.

adamziel replied:
I'd rather not introduce too many public-facing selectors in this PR. Let's take it into a follow-up one

kevin940726 commented on packages/data/src/redux-store/metadata/test/selectors.js:
```diff
@@ -1,117 +1,329 @@
-	isResolving,
-} from '../selectors';
+jest.useRealTimers();
+jest.setTimeout( 1000000 );
```
Any idea why we need this? I believe maybe you forgot to delete this 😅 ?

adamziel replied:
I did forget to remove it – I needed it to use a breakpoint :D removed now, thanks

kevin940726 commented on packages/data/src/redux-store/metadata/test/selectors.js:
```diff
@@ -1,117 +1,329 @@
+};
+
+async function resolve( registry, selector ) {
+	const resolution = registry.resolveSelect( 'store' )[ selector ]();
+	try {
+		await resolution;
+	} catch ( e ) {}
```
Nit: styling suggestion.

```suggestion
	try {
		await registry.resolveSelect( 'store' )[ selector ]();
	} catch ( error ) {}
```

kevin940726 commented on packages/data/src/redux-store/test/index.js:
```diff
@@ -235,3 +235,45 @@ describe( 'controls', () => {
+	it( 'resolves when the resolution succeeded', async () => {
+		shouldFail = false;
+		const promise = registry.resolveSelect( 'store' ).getItems();
+		jest.advanceTimersByTime( 1 );
```
This is oddly specific, maybe we could do `jest.runAllTimers()` or `jest.runAllTicks()`?

kevin940726 commented on packages/data/src/resolvers-cache-middleware.js:
```diff
@@ -41,7 +42,8 @@ const createResolversCacheMiddleware = ( registry, reducerKey ) => () => (
 				if (
-					value !== false ||
+					value?.status === Status.RESOLVING ||
+					! value?.status ||
```
We should also update the comments above.

kevin940726 reviewed (commented):
I like this a lot! Thank you for working on this, as always!

Left some minor feedback, but generally it looks good!

adamziel committed:
Rename e to error

adamziel committed:
Use undefined for an idle status so that getResolutionState can always return StateValue

adamziel committed:
Code style: Inline the awaited statement in resolve() in tests

adamziel committed:
Replace advanceTimersByTime with runAllTimers

adamziel committed:
Update the documentation and condition formulation in the resolve-cache-middleware

getdave commented:
> A promise returned by the resolveSelect will now be rejected in case of resolution error. Previously it just hanged forever.

This is _the best_ and it's [definitely required](https://github.com/WordPress/gutenberg/pull/38858). Thank you for working on this 👏

adamziel committed:
Remove the undefined status type and only use the values actually returned by the reducer

adamziel committed:
Simplify hasStartedResolution

adamziel commented:
Do you see any blockers @youknowriad, or are we good?

youknowriad reviewed (approved):
This looks good to me. Do we have any docs about the resolution/resolvers that need to be updated?

adamziel committed:
Add a changelog entry

adamziel commented:
@youknowriad thanks! I added a changelog entry. Other than that I don't think the meta store is documented anywhere. I grepped all the meta selector and meta action names and didn't find any markdown documentation other than the autogenerated ones. I will make sure to include that in the new documents that I'm writing.

Oh, and I wasn't sure about the dev note label but I added it anyway – let's figure it out when the time comes.
